### PR TITLE
Custom mint module impl

### PR DESCRIPTION
**Pull request by philipsu522**
### Description
Implement custom minting logic so that token release can be controlled by unix time (via x/epoch module). This is pretty similar to how Osmosis does it. Mainly, we will specify the following:
- `genesis_epoch_provisions` - total sei to add from genesis
- `reduction_period_in_epochs` - number of epochs before reducing rewards
- `reduction_factor` - fraction to reduce subsequent rewards by

It also removes old params like inflation, inflation_max...etc

### Testing
- Unit tests
- Verified genesis file:
```
238     "mint": {
239       "minter": {
240         "epoch_provisions": "0.000000000000000000"
241       },
242       "params": {
243         "mint_denom": "usei",
244         "genesis_epoch_provisions": "5000000.000000000000000000",
245         "reduction_period_in_epochs": "2",
246         "reduction_factor": "0.500000000000000000"
247       },
248       "halven_started_epoch": "0"
249     },
```
- Ran local chain with above parameters:
```
> cat ../mint_logs.txt| grep "minted\|epoch" | awk '{print $2, $3, $4, $5, $6, $7, $8, $9}' | uniq -c
   1 INF Current epoch 0
  12 INF Current epoch 1
  11 INF Current epoch 2
   1 INF minted coins from module account amount=2500000usei from=mint
  12 INF Current epoch 3
  12 INF Current epoch 4
   1 INF minted coins from module account amount=1250000usei from=mint
  12 INF Current epoch 5
  12 INF Current epoch 6
   1 INF minted coins from module account amount=625000usei from=mint
  12 INF Current epoch 7
  12 INF Current epoch 8
   1 INF minted coins from module account amount=312500usei from=mint
  12 INF Current epoch 9
  12 INF Current epoch 10
   1 INF minted coins from module account amount=156250usei from=mint
  12 INF Current epoch 11
  12 INF Current epoch 12
   1 INF minted coins from module account amount=78125usei from=mint
   3 INF Current epoch 13
```
- Queried total supply, ensured they added up to start + epoch provisions (property of geometric series)
```
> seid q bank total
pagination:
  next_key: null
  total: "0"
supply:
- amount: "100000000000004687500"
  denom: usei
> seid q mint epoch-provisions
312500.000000000000000000
```
- Ran with different configs incl:
  - different epoch provisions
  - different reduction rate
  - different epoch intervals

### CODE REVIEW

File: x/epoch/types/hooks.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+func NewMultiEpochHooks(hooks ...EpochHooks) MultiEpochHooks {
+	return hooks
+}
+
+// AfterEpochEnd is called when epoch is going to be ended, epochNumber is the number of epoch that is ending.
+func (h MultiEpochHooks) AfterEpochEnd(ctx sdk.Context, epoch Epoch) {
+	for i := range h {
+		cacheCtx, write := ctx.CacheContext()
+		h[i].AfterEpochEnd(cacheCtx, epoch)
+		write()
```
Comment (codchen) [inline x/epoch/types/hooks.go]: cache would not be flushed only if `AfterEpochEnd` panicks, which would halt the chain anyway before `Commit`, so we might not need CacheContext here?

File: x/mint/keeper/keeper.go [hunk 2]
```diff
@@ -16,14 +16,16 @@ type Keeper struct {
 	paramSpace       paramtypes.Subspace
 	stakingKeeper    types.StakingKeeper
 	bankKeeper       types.BankKeeper
+	epochKeeper      types.EpochKeeper
```
Comment (codchen) [inline x/mint/keeper/keeper.go]: is this used?

**Review by codchen — APPROVED** [review]

**Comment by codchen** [comment]
One caveat with the current approach is that it shifts the actual logic of minting from `mint` module to `epoch` module, so we need to be careful about `BeginBlock` ordering in app constructor. `epoch` now needs to go before modules that depend on minting results (distribution? not sure if there is any)

File: x/epoch/keeper/keeper.go [hunk 3]
```diff
@@ -39,6 +40,17 @@ func NewKeeper(
 	}
 }
 
+// Set the gamm hooks.
```
Comment (LCyson) [inline x/epoch/keeper/keeper.go]: nit: wrong comment?

File: x/epoch/types/hooks.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+type EpochHooks interface {
+	// the first block whose timestamp is after the duration is counted as the end of the epoch
+	AfterEpochEnd(ctx sdk.Context, epoch Epoch)
+	// new epoch is next block of epoch end block
+	BeforeEpochStart(ctx sdk.Context, epoch Epoch)
+}
+
+var _ EpochHooks = MultiEpochHooks{}
+
+// combine multiple gamm hooks, all hook functions are run in array sequence.
```
Comment (LCyson) [inline x/epoch/types/hooks.go]: ditto: gamm is a osmosis specific pool type

File: x/epoch/types/hooks.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+type MultiEpochHooks []EpochHooks
+
+func NewMultiEpochHooks(hooks ...EpochHooks) MultiEpochHooks {
+	return hooks
+}
+
+// AfterEpochEnd is called when epoch is going to be ended, epochNumber is the number of epoch that is ending.
+func (h MultiEpochHooks) AfterEpochEnd(ctx sdk.Context, epoch Epoch) {
+	for i := range h {
+		cacheCtx, write := ctx.CacheContext()
```
Comment (LCyson) [inline x/epoch/types/hooks.go]: are the `AfterEpochEnd` and `BeforeEpochStart` supposed to break in any circumstance? i.e. when we do cacheContext and write() if the mint module's afterEpochEnd call fails somehow will the chain runs into a bad state indefinitely?

**Review by LCyson — APPROVED** [review]

File: x/epoch/types/hooks.go (hunk 1 above)
Comment (philipsu522) [reply x/epoch/types/hooks.go]: good point, since we just panic i think it's fine to use original ctx

File: x/mint/keeper/keeper.go (hunk 2 above)
Comment (philipsu522) [reply x/mint/keeper/keeper.go]: nope, removed

File: x/epoch/types/hooks.go (hunk 5 above)
Comment (philipsu522) [reply x/epoch/types/hooks.go]: right now, only AfterEpochEnd calls anything. The logic will panic which will halt the chain so we can go in and fix things: https://github.com/sei-protocol/sei-chain/pull/124/files#diff-5d06e3a8167fa5a91ac1040e8e5da549aaa66942648a558e517c38ab01cf71a2R32

Comment (philipsu522) [reply x/epoch/types/hooks.go]: \cc @codchen we were discussing this in his comment as well

**Comment by philipsu522** [comment]
> One caveat with the current approach is that it shifts the actual logic of minting from `mint` module to `epoch` module, so we need to be careful about `BeginBlock` ordering in app constructor. `epoch` now needs to go before modules that depend on minting results (distribution? not sure if there is any)

good point. I don't think epoch has any dependencies so I put it first (since our x/upgrades module will depend on it too \cc @BrandonWeng)

**Pull request merged**